MaxText/train.py

```diff
@@ -167,6 +168,12 @@ def train_step(model, config, state, data, dropout_rng):
   rng1, gen_aqt_rng = jax.random.split(dropout_rng)
   aqt_rng, rng2 = jax.random.split(gen_aqt_rng)
 
+  if config.per_device_batch_size < 1:
+    # decimate data by proportion of per_device_batch_size
+    batch_size = max_utils.calculate_raw_input_batch_size(config.per_device_batch_size)
```

Comment:
Would it be better to store `batch_size` into `config` before calling `train_step` so that we don't have to keep calling `max_utils.calculate_raw_input_batch_size` to recalculate the value? I saw that `__setattr__` https://github.com/google/maxtext/blob/50f9300c4f4d7f6cbc52db538b62d0397262d663/MaxText/pyconfig.py#L148-L150 to modify `config` just raises an error. Is there a reason why `config` should not be modified?

Reply:
Yes that would be MUCH better. `config` can't be edited once given to a user but it can do stuff before it is used.

Generally you should calculate a `batch_to_load` and a `batch_to_train_on`.

Also there is no need for an `if` here. It is always safe to slice, right?

Reply:
Changed it so that `batch_size_to_load` and `batch_size_to_train_on` are calculated in `pyconfig.py` and then stored in `config`. Yes it should be safe to always slice but would it be better to only slice when necessary?